passive/dynamic_manifest.go

```diff
@@ -0,0 +1,222 @@
+package passive
+
+import (
+	"bytes"
+	"context"
+	"time"
+
+	"github.com/filecoin-project/go-f3/ec"
+	"github.com/filecoin-project/go-f3/gpbft"
+	"github.com/filecoin-project/go-f3/manifest"
+	logging "github.com/ipfs/go-log/v2"
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"go.uber.org/multierr"
+	"golang.org/x/xerrors"
+)
+
+var log = logging.Logger("f3-dynamic-manifest")
+
+const (
+	ManifestPubSubTopicName = "/f3/manifests/0.0.1"
+	ManifestCheckTick       = 5 * time.Second
+)
+
+var _ manifest.ManifestProvider = (*DynamicManifestProvider)(nil)
+
+// DynamicManifestProvider is a manifest provider that allows
+// the manifest to be changed at runtime.
+type DynamicManifestProvider struct {
+	manifest         manifest.Manifest
+	pubsub           *pubsub.PubSub
+	ec               ec.Backend
+	manifestServerID peer.ID
+
+	// these are populate in runtime
+	onManifestChange manifest.OnManifestChange
+	nextManifest     *manifest.Manifest
+	manifestTopic    *pubsub.Topic
+}
+
+func NewDynamicManifestProvider(manifest manifest.Manifest, pubsub *pubsub.PubSub, ec ec.Backend, manifestServerID peer.ID) manifest.ManifestProvider {
+	return &DynamicManifestProvider{
+		manifest:         manifest,
+		pubsub:           pubsub,
+		ec:               ec,
+		manifestServerID: manifestServerID,
+	}
+}
+
+func (m *DynamicManifestProvider) Manifest() manifest.Manifest {
+	return m.manifest
+}
+
+func (m *DynamicManifestProvider) GpbftOptions() []gpbft.Option {
+	return m.manifest.GpbftOptions()
+}
+
+func (m *DynamicManifestProvider) SetManifestChangeCallback(mc manifest.OnManifestChange) {
+	m.onManifestChange = mc
+}
+
+// When a manifest configuration changes, a new network name
+// is set that depends on the manifest version of the previous version to avoid
+// overlapping previous configurations.
+func (m *DynamicManifestProvider) networkNameOnChange() gpbft.NetworkName {
+	v, _ := m.manifest.Version()
+	return gpbft.NetworkName(string(m.manifest.NetworkName) + "/" + string(v))
+}
+
+func (m *DynamicManifestProvider) Run(ctx context.Context, errCh chan error) {
+	if m.onManifestChange == nil {
+		errCh <- xerrors.New("onManifestChange is nil. Callback for manifest change required")
+	}
+	go m.handleIncomingManifests(ctx, errCh)
+	m.handleApplyManifest(ctx, errCh)
+}
+
+func (m *DynamicManifestProvider) handleApplyManifest(ctx context.Context, errCh chan error) {
+	// add a timer for EC period
+	ticker := time.NewTicker(ManifestCheckTick)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			if m.nextManifest != nil {
+				ts, err := m.ec.GetHead(ctx)
+				if err != nil {
+					log.Errorf("error fetching chain head: %+v", err)
+					continue
+				}
+
+				// if the upgrade epoch is reached or already passed.
+				if ts.Epoch() >= m.nextManifest.BootstrapEpoch {
```

Comment:
What about the case where we didn't boostrap yet? we might want to swap out the manifest either way